### PR TITLE
k_mem_pool: Complete rework

### PR DESCRIPTION
**Pull request by andyross**
This patch amounts to a mostly complete rewrite of the k_mem_pool
allocator, which had been the source of historical complaints vs. the
one easily available in newlib.  The basic design of the allocator is
unchanged (it's still a 4-way buddy allocator), but the implementation
has made different choices throughout.  Major changes:

Space efficiency: The old implementation required ~2.66 bytes per
"smallest block" in overhead, plus 16 bytes per log4 "level" of the
allocation tree, plus a global tracking struct of 32 bytes and a very
surprising 12 byte overhead (in struct k_mem_block) per active
allocation on top of the returned data pointer.  This new allocator
uses a simple bit array as the only per-block storage and places the
free list into the freed blocks themselves, requiring only ~1.33 bits
per smallest block, 12 bytes per level, 32 byte globally and only 4
bytes of per-allocation bookeeping.  And it puts more of the generated
tree into BSS, slightly reducing binary sizes for non-trivial pool
sizes (even as the code size itself has increased a tiny bit).

IRQ safe: atomic operations on the store have been cut down to be at
most "4 bit sets and dlist operations" (i.e. a few dozen
instructions), reducing latency significantly and allowing us to lock
against interrupts cleanly from all APIs.  Allocations and frees can
be done from ISRs now without limitation (well, obviously you can't
sleep, so "timeout" must be K_NO_WAIT).

Deterministic performance: there is no more "defragmentation" step
that must be manually managed.  Block coalescing is done synchronously
at free time and takes constant time (strictly log4(num_levels)), as
the detection of four free "partner bits" is just a simple shift and
mask operation.

Cleaner behavior with odd sizes.  The old code assumed that the
specified maximum size would be a power of four multiple of the
minimum size, making use of non-standard buffer sizes problematic.
This implementation re-aligns the sub-blocks at each level and can
handle situations wehre alignment restrictions mean fewer than 4x will
be available.  If you want precise layout control, you can still
specify the sizes rigorously.  It just doesn't break if you don't.

More portable: the original implementation made use of GNU assembler
macros embedded inline within C __asm__ statements.  Not all
toolchains are actually backed by a GNU assembler even when the
support the GNU assembly syntax.  This is pure C, albeit with some
hairy macros to expand the compile-time-computed values.

Related changes that had to be rolled into this patch for bisectability:

* The new allocator has a firm minimum block size of 8 bytes (to store
  the dlist_node_t).  It will "work" with smaller requested min_size
  values, but obviously makes no firm promises about layout or how
  many will be available.  Unfortunately many of the tests were
  written with very small 4-byte minimum sizes and to assume exactly
  how many they could allocate.  Bump the sizes to match the allocator
  minimum.

* The mbox and pipes API made use of the internals of k_mem_block and
  had to be ported to the new scheme.  Blocks no longer store a
  backpointer to the pool that allocated them (it's an integer ID in a
  bitfield) , so if you want to "nullify" them you have to use the
  data pointer.

* test_mbox_api had a bug were it was prematurely freeing k_mem_blocks
  that it sent through the mailbox.  This worked in the old allocator
  because the memory wouldn't be touched when freed, but now we stuff
  list pointers in there and the bug was exposed.

* Remove test_mpool_options: the options (related to defragmentation
  behavior) tested no longer exist.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

File: include/kernel.h [hunk 1]
```diff
@@ -3079,199 +3088,80 @@ static inline u32_t k_mem_slab_num_free_get(struct k_mem_slab *slab)
...
-	u32_t nr_of_maxblocks;
-	u32_t nr_of_block_sets;
-	struct k_mem_pool_block_set *block_set;
-	char *bufblock;
+	void *buf;
+	size_t max_sz;
+	u16_t n_max;
+	u8_t n_levels;
+	u8_t max_inline_level;
+	struct k_mem_pool_lvl *levels;
```
Comment (lpereira) [inline include/kernel.h]: 1-byte hole between `max_inline_level` and `levels`.

File: kernel/mempool.c [hunk 2]
```diff
@@ -0,0 +1,404 @@
+#include <kernel.h>
```
Comment (lpereira) [inline kernel/mempool.c]: Missing copyright header.

File: kernel/mempool.c [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+int k_mem_pool_alloc(struct k_mem_pool *p, struct k_mem_block *block,
+		     size_t size, s32_t timeout)
+{
+	int ret, key;
+	s64_t end = 0;
+
+	__ASSERT(!(_is_in_isr() && timeout != K_NO_WAIT), "");
+
+	if (timeout > 0) {
+		end = _tick_get() + _ms_to_ticks(timeout);
```
Comment (lpereira) [inline kernel/mempool.c]: Unlikely but what happens if end overflows?

File: kernel/mempool.c [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+			clear_free_bit(p, level, b);
+			if (b != bn &&
+			    block_fits(p, block_ptr(p, lsz, b), lsz)) {
+				sys_dlist_remove(block_ptr(p, lsz, b));
+			}
+		}
+
+		irq_unlock(key);
+		free_block(p, level-1, lsizes, bn / 4); /* tail recursion! */
```
Comment (lpereira) [inline kernel/mempool.c]: Curious: Does the compiler really eliminates the call in this case?  I've seen it doing on cases where you return the result of calling a function recursively, transforming it into a JMP.

**Review by lpereira — CHANGES_REQUESTED** [review]

**Review by dbkinder — DISMISSED** [review]
+1 for the docs

File: include/kernel.h (hunk 1 above)
Comment (tbursztyka) [inline include/kernel.h]: Does not seem so ( 4/4/2/1/1/4 )

Comment (andrewboie) [inline include/kernel.h]: Yeah this looks ok to me, I don't see any place where padding would be inserted..

File: include/kernel.h [hunk 5]
```diff
@@ -3346,16 +3238,13 @@ extern void k_mem_pool_free(struct k_mem_block *block);
...
- * need for the memory pool to perform an automatic partial defragmentation.
+ * This is a no-op API preserved for backward compatibility only.
  *
- * @param pool Address of the memory pool.
+ * @param pool Unused
  *
  * @return N/A
  */
-extern void k_mem_pool_defrag(struct k_mem_pool *pool);
+static inline void __deprecated k_mem_pool_defrag(struct k_mem_pool *pool) {}
```
Comment (andrewboie) [inline include/kernel.h]: may need to put ARG_UNUSED(pool) here to avoid unused parameter warnings.

File: include/kernel.h (hunk 1 above)
Comment (lpereira) [inline include/kernel.h]: Yeah, it was a brain fart. Disregard my comment.

File: kernel/mempool.c (hunk 2 above)
Comment (andrewboie) [inline kernel/mempool.c]: it's fine, that;s what "SPDX-License-Identifier: Apache-2.0" does nowadays, we removed all the copyright boilerplate kernel-wide.

File: kernel/Makefile [hunk 6]
```diff
@@ -21,10 +21,9 @@ lib-y += $(strip \
 	queue.o \
 	stack.o \
 	mem_slab.o \
-	mem_pool.o \
+	mempool.o \
```
Comment (andrewboie) [inline kernel/Makefile]: any particular reason for the file rename?

File: kernel/mempool.c [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+
+	return bn & 0x1f;
+}
+
+static void set_free_bit(struct k_mem_pool *p, int level, int bn)
+{
+	u32_t *word;
+	int bit = get_bit_ptr(p, level, bn, &word);
+
+	*word |= (1<<bit);
```
Comment (andrewboie) [inline kernel/mempool.c]: strictly speaking, I think our coding style encourages spaces around the "<<" operator

File: kernel/mempool.c [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+}
+
+static void clear_free_bit(struct k_mem_pool *p, int level, int bn)
+{
+	u32_t *word;
+	int bit = get_bit_ptr(p, level, bn, &word);
+
+	*word &= ~(1<<bit);
```
Comment (andrewboie) [inline kernel/mempool.c]: see above

File: kernel/mempool.c (hunk 2 above)
Comment (lpereira) [inline kernel/mempool.c]: This file doesn't have the SPDX identifier.

Comment (andyross) [inline kernel/mempool.c]: For clarity: that tag was missing when Leandro left the comment.  I added it in an updated pull request.  Github isn't terribly clear about showing when that happens.

File: kernel/Makefile (hunk 6 above)
Comment (andyross) [inline kernel/Makefile]: Simply for clarity of review.  With the exception of k_malloc()/k_free (which are simple wrappers) all the code is new.  Easier to read that in a single chunk than a bunch of synthesized diffs.

**Pull request merged**